File: Frontend/static/js/health_profile.js

/* =====================================================
   HEALTH PROFILE JAVASCRIPT
   Backend API:
   GET  /health-profile
   POST /health-profile
   PUT  /health-profile
   ===================================================== */


const API_URL = "/health-profile";


const form =
    document.getElementById("healthForm");

const saveBtn =
    document.getElementById("saveBtn");

const messageBox =
    document.getElementById("messageBox");

const statusPill =
    document.getElementById("statusPill");

const lastUpdated =
    document.getElementById("lastUpdated");


let profileExists = false;


/*
=========================================================
Function Name: showMessage

Description:
Displays success or error message to the user.

Input:
message - Message text
isError - true for error, false for success

Output:
None
=========================================================
*/

function showMessage(message, isError = false) {

    messageBox.textContent = message;

    messageBox.classList.remove(
        "hidden",
        "error"
    );


    if (isError) {

        messageBox.classList.add("error");

    }

}


/*
=========================================================
Function Name: clearMessages

Description:
Clears previous messages and validation errors.

Input:
None

Output:
None
=========================================================
*/

function clearMessages() {

    messageBox.classList.add("hidden");

    messageBox.textContent = "";


    document
        .querySelectorAll(".error")
        .forEach(error => {

            error.textContent = "";

        });

}


/*
=========================================================
Function Name: getFormData

Description:
Collects health profile values from the form.

Input:
Form field values

Output:
Object containing backend-compatible health data
=========================================================
*/

function getFormData() {

    return {

        age:
            document.getElementById("age").value || "",

        gender:
            document.getElementById("gender").value || "",

        height_cm:
            document.getElementById("height_cm").value || "",

        weight_kg:
            document.getElementById("weight_kg").value || "",

        activity_level:
            document.getElementById(
                "activity_level"
            ).value || "",

        diet_preference:
            document.getElementById(
                "diet_preference"
            ).value || "",

        allergies:
            document.getElementById(
                "allergies"
            ).value || "",

        existing_conditions:
            document.getElementById(
                "existing_conditions"
            ).value || ""

    };

}


/*
=========================================================
Function Name: fillForm

Description:
Displays database health profile data in the form.

Input:
Profile object returned by backend

Output:
None
=========================================================
*/

function fillForm(profile) {

    document.getElementById("age").value =
        profile.age ?? "";


    document.getElementById("gender").value =
        profile.gender ?? "";


    document.getElementById("height_cm").value =
        profile.height_cm ?? "";


    document.getElementById("weight_kg").value =
        profile.weight_kg ?? "";


    document.getElementById("activity_level").value =
        profile.activity_level ?? "sedentary";


    document.getElementById("diet_preference").value =
        profile.diet_preference ?? "";


    document.getElementById("allergies").value =
        profile.allergies ?? "";


    document.getElementById(
        "existing_conditions"
    ).value =
        profile.existing_conditions ?? "";


    updateBMI();


    if (profile.updated_at) {

        lastUpdated.textContent =
            "Last updated: " +
            profile.updated_at;

    }

}


/*
=========================================================
Function Name: displayValidationErrors

Description:
Displays validation errors returned by backend.

Input:
errors - Backend validation error object

Output:
None
=========================================================
*/

function displayValidationErrors(errors) {

    Object.entries(errors || {})
        .forEach(([field, message]) => {

            const errorElement =
                document.getElementById(
                    `${field}Error`
                );


            if (errorElement) {

                errorElement.textContent =
                    message;

            }

        });

}


/*
=========================================================
Function Name: calculateBMI

Description:
Calculates BMI using height and weight.

Input:
height_cm - Height in centimeters
weight_kg - Weight in kilograms

Output:
BMI value or null
=========================================================
*/

function calculateBMI(
    height_cm,
    weight_kg
) {

    const height =
        Number(height_cm);

    const weight =
        Number(weight_kg);


    if (
        !height ||
        !weight ||
        height <= 0 ||
        weight <= 0
    ) {

        return null;

    }


    const heightMeters =
        height / 100;


    return (
        weight /
        (heightMeters * heightMeters)
    );

}


/*
=========================================================
Function Name: getBMILabel

Description:
Returns BMI category.

Input:
bmi - BMI value

Output:
BMI category
=========================================================
*/

function getBMILabel(bmi) {

    if (bmi < 18.5) {

        return "Underweight";

    }

    if (bmi < 25) {

        return "Normal";

    }

    if (bmi < 30) {

        return "Overweight";

    }

    return "Obesity";

}


/*
=========================================================
Function Name: updateBMI

Description:
Updates BMI displayed on the page.

Input:
Current height and weight

Output:
None
=========================================================
*/

function updateBMI() {

    const bmi =
        calculateBMI(

            document.getElementById(
                "height_cm"
            ).value,

            document.getElementById(
                "weight_kg"
            ).value

        );


    const bmiValue =
        document.getElementById(
            "bmiValue"
        );


    const bmiLabel =
        document.getElementById(
            "bmiLabel"
        );


    const sideBmi =
        document.getElementById(
            "sideBmi"
        );


    if (bmi === null) {

        bmiValue.textContent = "--";

        bmiLabel.textContent =
            "Enter height & weight";

        sideBmi.textContent = "--";

        return;

    }


    const rounded =
        bmi.toFixed(1);


    bmiValue.textContent =
        rounded;


    bmiLabel.textContent =
        getBMILabel(bmi);


    sideBmi.textContent =
        rounded;

}


/*
=========================================================
Function Name: loadHealthProfile

Description:
Fetches logged-in user's health profile
from Flask backend.

Input:
None

Output:
Loads profile data into form
=========================================================
*/

async function loadHealthProfile() {

    try {

        statusPill.textContent =
            "Loading...";


        const response =
            await fetch(
                API_URL,
                {
                    method: "GET",

                    credentials:
                        "same-origin"
                }
            );


        const result =
            await response.json();


        /*
        USER NOT LOGGED IN
        */

        if (response.status === 401) {

            statusPill.textContent =
                "Login required";


            showMessage(
                result.message ||
                "Please login first.",
                true
            );

            return;

        }


        /*
        OTHER BACKEND ERROR
        */

        if (
            !response.ok ||
            !result.success
        ) {

            statusPill.textContent =
                "Error";


            showMessage(
                result.message ||
                "Unable to load profile.",
                true
            );

            return;

        }


        /*
        PROFILE EXISTS
        */

        profileExists =
            result.profile_exists === true;


        if (
            profileExists &&
            result.profile
        ) {

            fillForm(
                result.profile
            );


            statusPill.textContent =
                "Profile saved";

        }


        /*
        PROFILE DOES NOT EXIST
        */

        else {

            document.getElementById(
                "activity_level"
            ).value =
                "sedentary";


            updateBMI();


            statusPill.textContent =
                "New profile";

        }

    }

    catch (error) {

        console.error(error);


        statusPill.textContent =
            "Connection error";


        showMessage(
            "Unable to connect to Health Profile API.",
            true
        );

    }

}


/*
=========================================================
Function Name: saveHealthProfile

Description:
Creates or updates health profile.

Input:
Health profile form data

Output:
Backend success/error message
=========================================================
*/

async function saveHealthProfile() {

    clearMessages();


    const data =
        getFormData();


    saveBtn.disabled = true;

    saveBtn.textContent =
        "Saving...";


    /*
    If profile exists -> PUT

    If profile doesn't exist -> POST
    */

    const method =
        profileExists
            ? "PUT"
            : "POST";


    try {

        const response =
            await fetch(
                API_URL,
                {

                    method: method,

                    headers: {

                        "Content-Type":
                            "application/json"

                    },

                    credentials:
                        "same-origin",

                    body:
                        JSON.stringify(data)

                }
            );


        const result =
            await response.json();


        /*
        LOGIN CHECK
        */

        if (response.status === 401) {

            showMessage(
                result.message ||
                "Please login first.",
                true
            );

            return;

        }


        /*
        VALIDATION ERROR
        */

        if (response.status === 400) {

            displayValidationErrors(
                result.errors
            );


            showMessage(
                "Please correct the highlighted fields.",
                true
            );

            return;

        }


        /*
        PROFILE ALREADY EXISTS
        */

        if (response.status === 409) {

            profileExists = true;


            showMessage(
                result.message ||
                "Health profile already exists.",
                true
            );

            return;

        }


        /*
        PROFILE NOT FOUND
        */

        if (response.status === 404) {

            profileExists = false;


            showMessage(
                result.message ||
                "Health profile not found.",
                true
            );

            return;

        }


        /*
        OTHER ERROR
        */

        if (
            !response.ok ||
            !result.success
        ) {

            showMessage(
                result.message ||
                "Unable to save profile.",
                true
            );

            return;

        }


        /*
        SUCCESS
        */

        profileExists = true;


        statusPill.textContent =
            "Profile saved";


        showMessage(
            result.message ||
            "Health profile saved successfully."
        );


        /*
        Reload data from database
        */

        await loadHealthProfile();

    }

    catch (error) {

        console.error(error);


        showMessage(
            "Unable to connect to backend.",
            true
        );

    }

    finally {

        saveBtn.disabled = false;

        saveBtn.textContent =
            "Save Changes";

    }

}


/*
=========================================================
FORM SUBMIT
=========================================================
*/

form.addEventListener(
    "submit",
    function(event) {

        event.preventDefault();

        saveHealthProfile();

    }
);


/*
=========================================================
LIVE BMI
=========================================================
*/

document
    .getElementById("height_cm")
    .addEventListener(
        "input",
        updateBMI
    );


document
    .getElementById("weight_kg")
    .addEventListener(
        "input",
        updateBMI
    );


/*
=========================================================
TAB NAVIGATION
=========================================================
*/

document
    .querySelectorAll(".tab")
    .forEach(tab => {

        tab.addEventListener(
            "click",
            function() {

                document
                    .querySelectorAll(".tab")
                    .forEach(t => {

                        t.classList.remove(
                            "active"
                        );

                    });


                this.classList.add(
                    "active"
                );


                const section =
                    document.getElementById(
                        this.dataset.section
                    );


                if (section) {

                    section.scrollIntoView({
                        behavior: "smooth",
                        block: "start"
                    });

                }

            }
        );

    });


/*
=========================================================
LOAD PROFILE WHEN PAGE OPENS
=========================================================
*/

loadHealthProfile();

File: Backend/health_profile.py
from Database_Module import connection
from flask import Blueprint, request, jsonify, session
import mysql.connector
from mysql.connector import Error


# Create Health Profile Blueprint
health_bp = Blueprint("health", __name__)


# =========================================================
# VALIDATION FUNCTION
# =========================================================
"""
    Function Name: validate_health_profile
    Description: Validates health profile fields such as age, gender, height,
                 weight, and activity level.
    Input: data (dictionary containing health profile information)
    Output: errors (dictionary containing validation errors)
"""
def validate_health_profile(data):

    errors = {}

    # -------------------------
    # Age validation
    # -------------------------
    age = data.get("age")

    if age is not None and age != "":
        try:
            age = int(age)

            if age <= 0 or age > 120:
                errors["age"] = "Age must be between 1 and 120."

        except (ValueError, TypeError):
            errors["age"] = "Age must be a valid number."


    # -------------------------
    # Gender validation
    # -------------------------
    gender = data.get("gender")

    allowed_gender = [
        "male",
        "female",
        "other"
    ]

    if gender not in [None, ""]:

        if gender not in allowed_gender:
            errors["gender"] = "Invalid gender."


    # -------------------------
    # Height validation
    # -------------------------
    height = data.get("height_cm")

    if height is not None and height != "":
        try:
            height = float(height)

            if height <= 0:
                errors["height_cm"] = "Height must be greater than 0."

        except (ValueError, TypeError):
            errors["height_cm"] = "Height must be a valid number."


    # -------------------------
    # Weight validation
    # -------------------------
    weight = data.get("weight_kg")

    if weight is not None and weight != "":
        try:
            weight = float(weight)

            if weight <= 0:
                errors["weight_kg"] = "Weight must be greater than 0."

        except (ValueError, TypeError):
            errors["weight_kg"] = "Weight must be a valid number."


    # -------------------------
    # Activity level validation
    # -------------------------
    activity_level = data.get("activity_level")

    allowed_activity = [
        "sedentary",
        "light",
        "moderate",
        "active"
    ]

    if activity_level not in [None, ""]:

        if activity_level not in allowed_activity:
            errors["activity_level"] = "Invalid activity level."


    # Return validation errors
    return errors


# =========================================================
# GET HEALTH PROFILE
# =========================================================
"""
    Function Name: get_health_profile
    Description: Fetches the logged-in user's health profile from the database.
    Input: Logged-in user's session user_id
    Output: JSON response containing the health profile or an error message
"""
@health_bp.route("/health-profile", methods=["GET"])
def get_health_profile():

    db = None
    cursor = None

    # Check whether user is logged in
    if "user_id" not in session:

        return jsonify({
            "success": False,
            "message": "Please login first."
        }), 401

    user_id = session["user_id"]

    try:

        db = connection()
        cursor = db.cursor(dictionary=True)

        query = """
            SELECT
                profile_id,
                user_id,
                age,
                gender,
                height_cm,
                weight_kg,
                activity_level,
                diet_preference,
                allergies,
                existing_conditions,
                updated_at
            FROM health_profiles
            WHERE user_id = %s
        """

        cursor.execute(query, (user_id,))

        profile = cursor.fetchone()

        # Profile does not exist
        if profile is None:

            return jsonify({
                "success": True,
                "profile_exists": False,
                "profile": None
            }), 200

        # Profile found
        return jsonify({
            "success": True,
            "profile_exists": True,
            "profile": profile
        }), 200

    except Error as e:

        return jsonify({
            "success": False,
            "message": "Database error occurred."
        }), 500

    finally:

        if cursor:
            cursor.close()

        if db:
            db.close()

# =========================================================
# CREATE HEALTH PROFILE
# =========================================================
"""
    Function Name: create_health_profile
    Description: Validates and creates a new health profile for the logged-in user.
    Input: JSON health profile data from the request
    Output: JSON response with success message and profile_id
"""
@health_bp.route("/health-profile", methods=["POST"])
def create_health_profile():

    db = None
    cursor = None

    # Check login
    if "user_id" not in session:

        return jsonify({
            "success": False,
            "message": "Please login first."
        }), 401

    # Get logged-in user's ID
    user_id = session["user_id"]

    # Get JSON data
    data = request.get_json(silent=True)

    if not data:

        return jsonify({
            "success": False,
            "message": "No health profile data received."
        }), 400

    # Validate data
    errors = validate_health_profile(data)

    if errors:

        return jsonify({
            "success": False,
            "errors": errors
        }), 400

    try:

        db = connection()
        cursor = db.cursor(dictionary=True)

        # -------------------------------------------------
        # Check whether profile already exists
        # -------------------------------------------------

        check_query = """
            SELECT profile_id
            FROM health_profiles
            WHERE user_id = %s
        """

        cursor.execute(check_query, (user_id,))

        existing_profile = cursor.fetchone()

        if existing_profile:

            return jsonify({
                "success": False,
                "message": "Health profile already exists. Use PUT to update it."
            }), 409

        # -------------------------------------------------
        # Insert new profile
        # -------------------------------------------------

        insert_query = """
            INSERT INTO health_profiles
            (
                user_id,
                age,
                gender,
                height_cm,
                weight_kg,
                activity_level,
                diet_preference,
                allergies,
                existing_conditions
            )
            VALUES
            (
                %s, %s, %s, %s, %s,
                %s, %s, %s, %s
            )
        """

        values = (
            user_id,
            data.get("age") or None,
            data.get("gender") or None,
            data.get("height_cm") or None,
            data.get("weight_kg") or None,
            data.get("activity_level") or "sedentary",
            data.get("diet_preference") or None,
            data.get("allergies") or None,
            data.get("existing_conditions") or None
        )

        cursor.execute(insert_query, values)

        db.commit()

        return jsonify({
            "success": True,
            "message": "Health profile created successfully.",
            "profile_id": cursor.lastrowid
        }), 201

    except Error as e:

        if db:
            db.rollback()

        return jsonify({
            "success": False,
            "message": "Unable to create health profile."
        }), 500

    finally:

        if cursor:
            cursor.close()

        if db:
            db.close()

# =========================================================
# UPDATE HEALTH PROFILE
# =========================================================
"""
    Function Name: update_health_profile
    Description: Validates and updates the existing health profile of the logged-in user.
    Input: JSON health profile data from the request
    Output: JSON response with update status and message
"""
@health_bp.route("/health-profile", methods=["PUT"])
def update_health_profile():

    db = None
    cursor = None

    # Check login
    if "user_id" not in session:

        return jsonify({
            "success": False,
            "message": "Please login first."
        }), 401

    user_id = session["user_id"]

    # Get JSON data
    data = request.get_json(silent=True)

    if not data:

        return jsonify({
            "success": False,
            "message": "No health profile data received."
        }), 400

    # Validate data
    errors = validate_health_profile(data)

    if errors:

        return jsonify({
            "success": False,
            "errors": errors
        }), 400

    try:

        db = connection()
        cursor = db.cursor(dictionary=True)

        # -------------------------------------------------
        # Check profile exists
        # -------------------------------------------------

        check_query = """
            SELECT profile_id
            FROM health_profiles
            WHERE user_id = %s
        """

        cursor.execute(check_query, (user_id,))

        existing_profile = cursor.fetchone()

        if not existing_profile:

            return jsonify({
                "success": False,
                "message": "Health profile not found."
            }), 404

        # -------------------------------------------------
        # Update profile
        # -------------------------------------------------

        update_query = """
            UPDATE health_profiles
            SET
                age = %s,
                gender = %s,
                height_cm = %s,
                weight_kg = %s,
                activity_level = %s,
                diet_preference = %s,
                allergies = %s,
                existing_conditions = %s,
                updated_at = CURRENT_TIMESTAMP
            WHERE user_id = %s
        """

        values = (
            data.get("age") or None,
            data.get("gender") or None,
            data.get("height_cm") or None,
            data.get("weight_kg") or None,
            data.get("activity_level") or "sedentary",
            data.get("diet_preference") or None,
            data.get("allergies") or None,
            data.get("existing_conditions") or None,
            user_id
        )

        cursor.execute(update_query, values)

        db.commit()

        return jsonify({
            "success": True,
            "message": "Health profile updated successfully."
        }), 200

    except Error as e:

        if db:
            db.rollback()

        return jsonify({
            "success": False,
            "message": "Unable to update health profile."
        }), 500

    finally:

        if cursor:
            cursor.close()

        if db:
            db.close()